File: day_2/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

int main() {
    FILE *file = fopen("../input.txt", "r");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    char line[65536]; // Assuming the line fits in buffer
    if (!fgets(line, sizeof(line), file)) {
        fclose(file);
        return 1;
    }
    fclose(file);

    // Remove trailing newline if present
    line[strcspn(line, "\r\n")] = 0;

    long long total_invalid = 0;
    char buffer[64]; // Buffer for number string conversion

    struct timespec start_time, end_time;
    clock_gettime(CLOCK_MONOTONIC, &start_time);

    char *id_range = strtok(line, ",");
    while (id_range != NULL) {
        long start, end;
        char *dash = strchr(id_range, '-');
        if (dash) {
            *dash = '\0';
            start = atol(id_range);
            end = atol(dash + 1);
            
            for (long number = start; number <= end; number++) {
                // Convert number to string
                int len = sprintf(buffer, "%ld", number);
                
                if (len % 2 != 0) {
                    continue;
                }
                
                int half = len / 2;
                // Compare first half and second half
                // first half is buffer[0]...buffer[half-1]
                // second half is buffer[half]...buffer[len-1]
                
                if (strncmp(buffer, buffer + half, half) == 0) {
                    total_invalid += number;
                }
            }
        }
        id_range = strtok(NULL, ",");
    }

    clock_gettime(CLOCK_MONOTONIC, &end_time);
    double time_taken = (end_time.tv_sec - start_time.tv_sec) +
                        (end_time.tv_nsec - start_time.tv_nsec) / 1e9;

    printf("Execution time: %.6f seconds\n", time_taken);
    printf("Total invalid ID count: %lld\n", total_invalid);

    return 0;
}

File: day_2/rust/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::time::Instant;

fn main() -> io::Result<()> {
    // Attempt to open the file. Assuming running from day_2/rust/
    let path = Path::new("../input.txt");
    let file = File::open(&path)?;
    let mut reader = io::BufReader::new(file);

    let mut line = String::new();
    reader.read_line(&mut line)?;
    let line = line.trim();

    let mut total_invalid = 0;

    let start_time = Instant::now();

    for id_range in line.split(',') {
        let parts: Vec<&str> = id_range.split('-').collect();
        if parts.len() != 2 {
            continue;
        }
        let start: i64 = parts[0].parse().unwrap();
        let end: i64 = parts[1].parse().unwrap();

        for number in start..=end {
            let s = number.to_string();
            let len = s.len();
            if len % 2 != 0 {
                continue;
            }
            let half = len / 2;
            let first_half = &s[..half];
            let second_half = &s[half..];

            if first_half == second_half {
                total_invalid += number;
            }
        }
    }

    let duration = start_time.elapsed();
    println!("Execution time: {:.6} seconds", duration.as_secs_f64());
    println!("Total invalid ID count: {}", total_invalid);

    Ok(())
}